doc/synchronization.md

```diff
@@ -0,0 +1,356 @@
+that all instance variables that are set only once in constructor (therefore
+effectively final) are safely published to all readers (assuming proper
+construction - object instance does not escape during construction).
+
+``` ruby
+class ImmutableTreeNode < Concurrent::Synchronization::Object
+  # mark this class to publish final instance variables safely
+  safe_initialization!
+
+  def initialize(left, right)
+    # Call super to allow proper initialization.
+    super()
+    # By convention final instance variables have CamelCase names
+    # to distinguish them from ordinary instance variables.
+    @Left  = left
+    @Right = right
+  end
+
+  # Define thread-safe readers.
+  def left
+    # No need to synchronize or otherwise protect, it's already
+    # guaranteed to be visible.
+    @Left
+  end    
+
+  def right
+    @Right
+  end
+end
+```
+
+Once `safe_initialization!` is called on a class it transitively applies to all
+its children.
+
+> It's implemented by adding `new`, when `safe_initialization!` is called, as
+> follows:
+>
+> ``` ruby
+> def self.new(*)
+>   object = super
+> ensure
+>   object.ensure_ivar_visibility! if object
+> end
+> ```
+>
+> therefore `new` should not be overridden.
+
+## Volatile instance variable
+
+`Synchronization::Object` children can have volatile instance variables. A Ruby
+library cannot alter meaning of `@a_name` expression therefore when a
+`attr_volatile :a_name` is called, declaring the instance variable named
+`a_name` to be volatile, it creates method accessors.
+
+> However there is Ruby [issue](https://redmine.ruby-lang.org/issues/11539)
+> filled to address this.
+
+``` ruby
+# Simple counter with cheap reads.
+class Counter < Concurrent::Synchronization::Object
+  # Declare instance variable value to be volatile and its
+  # reader and writer to be private. `attr_volatile` returns
+  # names of created methods.
+  private *attr_volatile(:value)
+  safe_initialization!
+
+  def initialize(value)
+    # Call super to allow proper initialization.
+    super()
+    # Create a reentrant lock instance held in final ivar
+    # to be able to protect writer.
+    @Lock = Concurrent::Synchronization::Lock.new
+    # volatile write
+    self.value = value
+  end
+
+  # Very cheap reader of the Counter's current value, just a volatile read.
+  def count
+    # volatile read
+    value
+  end
+
+  # Safely increment the value without loosing updates
+  # (as it would happen with just += used).
+  def increment(add)
+    # Wrap he two volatile operations to make them atomic.
+    @Lock.synchronize do
+      # volatile write and read
+      self.value = self.value + add
+    end
+  end   
+end
+```
+
+> This is currently planned to be migrated to a module to be able to add
+> volatile fields any object not just `Synchronization::Object` children. The
+> instance variable itself is named `@"volatile_#(name)"` to distinguish it and
```

Comment:
I don't think the quotes are necessary at all, the backticks provide enough highlighting.